source/common/network/address_impl.h

```diff
@@ -6,16 +6,74 @@
 #include <cstdint>
 #include <string>
 
+#include "envoy/common/exception.h"
 #include "envoy/common/platform.h"
 #include "envoy/network/address.h"
 #include "envoy/network/socket.h"
 
 #include "common/common/assert.h"
+#include "common/common/statusor.h"
+#include "common/common/thread.h"
 
 namespace Envoy {
 namespace Network {
 namespace Address {
 
+class InstanceFactory {
+public:
+  template <typename InstanceType, typename... Args>
+  static StatusOr<InstanceConstSharedPtr> createInstancePtr(Args&&... args) {
+    absl::Status status;
+    status = InstanceType::validateProtocolsupported();
+    if (!status.ok()) {
+      return status;
+    }
+    std::shared_ptr<InstanceType> instance(new InstanceType(std::forward<Args>(args)...));
+    if (!instance->initError().ok()) {
+      return instance->initError();
+    }
+    return instance;
+  }
+
+  template <typename InstanceType, typename... Args>
+  static InstanceConstSharedPtr createInstancePtrOrThrow(Args&&... args) {
+    StatusOr<InstanceConstSharedPtr> error_or_address =
+        InstanceFactory::createInstancePtr<InstanceType, Args...>(std::forward<Args>(args)...);
+    if (error_or_address.ok()) {
+      return *error_or_address;
+    } else {
+      ASSERT(Thread::MainThread::isMainThread());
+      throw EnvoyException(error_or_address.status().ToString());
```

Comment:
Put the ASSERT outside the if-statement.

Reply:
But some instance constructors never produce an error, can we allow them not to be checked in worker thread?

Reply:
In those cases, this function should not be called. Instead `createInstacePtr` should be called and status-checked.

Reply:
Makes sense.